### PR TITLE
Implement range support in `ocean.text.convert.Formatter`

### PR DESCRIPTION
**Pull request by MichaelKim20**
I Implement range support in `ocean.text.convert.Formatter`
Relates to bpfkorea/agora#718

### CODE REVIEW

File: src/ocean/text/convert/Formatter_test.d [hunk 1]
```diff
@@ -559,3 +559,23 @@ unittest
...
+{
+    bool[string] foo;
+    foo["A"] = true;
+    foo["B"] = true;
+    test(format("{}", foo.byKey()) == `[ "A", "B" ]`);
+
+    bool[int] bar;
+    bar[0] = true;
+    bar[1] = true;
+    test(format("{}", bar.byKey()) == `[ 0, 1 ]`);
```
Comment (AndrejMitrovic) [inline src/ocean/text/convert/Formatter_test.d]: How about adding a test where the range is empty?

File: src/ocean/text/convert/Formatter.d [hunk 2]
```diff
@@ -478,7 +478,30 @@ private void handle (T) (T v, FormatInfo f, scope FormatterSink sf, scope ElemSi
...
 
+            handle(v.front, f, sf, se);
+            v.popFront();
+        }
+        if (started)
+            sf(" ]");
+        else
+            sf("[]");
+        f.flags = old_flag;
+    }
```
Comment (AndrejMitrovic) [inline src/ocean/text/convert/Formatter.d]: I think we can simplify this a little bit:
```suggestion
        if (v.empty)
        {
            sf("[]");
            return;
        }
        
        Flags old_flag = f.flags;
        f.flags |= Flags.Nested;

        sf("[ ");
        while (1)
        {
            handle(v.front, f, sf, se);
            v.popFront();
            if (v.empty)
                break;
            else
                sf(", ");
        }
        sf(" ]");
        f.flags = old_flag;
    }
```

But to make sure this works I suggest adding two additional test-cases:
- Test with a range with 1 element
- Test with an empty range.

Comment (MichaelKim20) [reply src/ocean/text/convert/Formatter.d]: That's a good suggestion. They're all applied.

File: src/ocean/text/convert/Formatter.d [hunk 3]
```diff
@@ -962,3 +984,35 @@ unittest
     static assert(canSwitchOn!(immutable int));
     static assert(!canSwitchOn!(real));
 }
+
+private template isInputRange (R)
+{
+    enum bool isInputRange =
```
Comment (AndrejMitrovic) [inline src/ocean/text/convert/Formatter.d]: This is partially correct, but it's not checking the type of the return value of `empty` and `front`.

For example:

https://github.com/dlang/phobos/blob/8a38e6e56218923e04b623362daa4eb78f9eb442/std/range/primitives.d#L172-L177

Since `ReturnType` is complicated and requires copying a lot of code from Phobos, I suggest instead to rewrite `isInputRange` just slightly:

```D
enum bool isInputRange(R) =
    is(typeof(R.init) == R)
    && is(typeof((R r) { return r.empty; }(R.init)) == bool)
    && is(typeof((return ref R r) => r.front))
    && !is(typeof((R r) { return r.front; }(R.init)) == void)
    && is(typeof((R r) => r.popFront));
```

And also the two unittests from Phobos can be copied over:

```D

///
@safe unittest
{
    struct A {}
    struct B
    {
        void popFront();
        @property bool empty();
        @property int front();
    }

    static assert(!isInputRange!A);
    static assert( isInputRange!B);
    static assert( isInputRange!(int[]));
    static assert( isInputRange!(char[]));
    static assert(!isInputRange!(char[4]));
    static assert( isInputRange!(inout(int)[]));

    static struct NotDefaultConstructible
    {
        @disable this();
        void popFront();
        @property bool empty();
        @property int front();
    }
    static assert( isInputRange!NotDefaultConstructible);

    static struct NotDefaultConstructibleOrCopyable
    {
        @disable this();
        @disable this(this);
        void popFront();
        @property bool empty();
        @property int front();
    }
    static assert(isInputRange!NotDefaultConstructibleOrCopyable);

    static struct Frontless
    {
        void popFront();
        @property bool empty();
    }
    static assert(!isInputRange!Frontless);

    static struct VoidFront
    {
        void popFront();
        @property bool empty();
        void front();
    }
    static assert(!isInputRange!VoidFront);
}

@safe unittest
{
    import std.algorithm.comparison : equal;

    static struct R
    {
        static struct Front
        {
            R* impl;
            @property int value() { return impl._front; }
            alias value this;
        }

        int _front;

        @property bool empty() { return _front >= 3; }
        @property auto front() { return Front(&this); }
        void popFront() { _front++; }
    }
    R r;

    static assert(isInputRange!R);
    assert(r.equal([ 0, 1, 2 ]));
}
```

The tests pass with this.

@Geod24 I've just noticed something with my suggestion:

```D
static struct S
{
    @disable this();
    void popFront();
    @property bool empty();
    @property int front();

}

// == 'bool'. But how come 'S.init' doesn't error?
pragma(msg, typeof((S r) { return r.empty; }(S.init)));
```

It seems that some semantic checks are not done within the `typeof()` expression? I wonder if it's on purpose, and if so then `isInputRange` in Phobos could be simplified by getting rid of the `ReturnType` instantiation..

**Review by AndrejMitrovic — CHANGES_REQUESTED** [review]
I've requested changes.

File: src/ocean/text/convert/Formatter.d (hunk 3 above)
Comment (MichaelKim20) [reply src/ocean/text/convert/Formatter.d]: I also saw the source code and test code of Phobos, which was strange. It didn't work, so I wrote it using the least.

Comment (Geod24) [reply src/ocean/text/convert/Formatter.d]: @AndrejMitrovic : I'm not sure what you are referring to here. What error do you expect ?

Comment (MichaelKim20) [reply src/ocean/text/convert/Formatter.d]: Why not these?
``` D
    static assert( isInputRange!(int[]));
    static assert( isInputRange!(char[]));
    static assert(!isInputRange!(char[4]));
    static assert( isInputRange!(inout(int)[]));
```

Comment (AndrejMitrovic) [reply src/ocean/text/convert/Formatter.d]: Only `char[4]` is not an input range. The rest are input ranges.

`char[4]` is not an input range because you cannot call `popFront` on it.

**Comment by AndrejMitrovic** [comment]
The reason it fails to work in Ocean but works in Phobos is because the `front()` / `popFront` / `empty` methods are implemented in Phobos.

For example:

https://github.com/dlang/phobos/blob/227132664cfcb1ab5180f44eee287dfe253675bc/std/range/primitives.d#L2234

This uses UFCS, so if you do this:

```D
import std.range;
int[] arr;
arr.popFront()
```

The compiler rewrites this to:

```D
import std.range;
int[] arr;
popFront(arr);  // popFront implemented in std.range
```

You would need to implement `front`, `popFront`, `empty` in order to support arrays in Ocean.

@Geod24 do you have other suggestions?

**Comment by AndrejMitrovic** [comment]
@Geod24 ignore my previous comment, arrays are already handled.

**Comment by MichaelKim20** [comment]
> @Geod24 ignore my previous comment, arrays are already handled.

@AndrejMitrovic Thanks for your comment

File: src/ocean/text/convert/Formatter.d [hunk 4]
```diff
@@ -478,7 +478,29 @@ private void handle (T) (T v, FormatInfo f, scope FormatterSink sf, scope ElemSi
...
+        while (1)
+        {
+            handle(v.front, f, sf, se);
+            v.popFront();
+            if (v.empty)
+                break;
+            else
+                sf(", ");
+        }
+        sf(" ]");
```
Comment (Geod24) [inline src/ocean/text/convert/Formatter.d]:
```suggestion
        sf("[");
        while (!v.empty)
        {
            handle(v.front, f, sf, se);
            v.popFront();
            if (!v.empty)
                sf(", ");
        }
        sf("]");
```

And you can remove the special case for empty ranges

Comment (MichaelKim20) [reply src/ocean/text/convert/Formatter.d]: Oh, great. It's very simple, it's clear

**Comment by AndrejMitrovic** [comment]
I tested it with Agora:

> 2020-05-08 13:32:54,164 Info [agora.network.NetworkManager] - Discovering from ["GDGJ6LQ2WQHM4UPHPPF5UWVQFNUBG2UEI4PYKVIQO7TER3U3S7OIBTAV", "GAKU3XJPQPIDZKEQQW4GIIRH4COU7WUFYSNWUC5LPUE3MF3DV6D6CPV3"]

It seems to work great.

**Review by AndrejMitrovic — APPROVED** [review]

**Comment by Geod24** [comment]
Actually there's a few peculiarities I would like to think through before we merge this.
First, infinite range. Should we support them ?
Second, forward ranges. Should we restrict ourselves to forward ranges so as not to consume input ranges ?

**Comment by AndrejMitrovic** [comment]
Oh right, `.save` should ideally be used. Otherwise we consume the range.

As for infinite ranges, maybe we could just reject compilation? Then the user would have to do something like `log(infinite_range.take(5));` explicitly?

**Comment by MichaelKim20** [comment]
> Actually there's a few peculiarities I would like to think through before we merge this.
> First, infinite range. Should we support them ?
> Second, forward ranges. Should we restrict ourselves to forward ranges so as not to consume input ranges ?

@Geod24 Thank you for your good suggestion.
I'll think about it more more more deeply.

**Comment by MichaelKim20** [comment]
@Geod24 @AndrejMitrovic I am going to implement `isForwardRange` and `isInfinite`

**Comment by MichaelKim20** [comment]
I added `isForwardRange` and `isInfinite`.

And about `Infinity Range`,
I would apply it to `agora` in the way @AndrejMitrovic  suggested.

@Geod24 ping

File: src/ocean/text/convert/Formatter.d [hunk 5]
```diff
@@ -478,7 +478,21 @@ private void handle (T) (T v, FormatInfo f, scope FormatterSink sf, scope ElemSi
         nullWrapper(&v, se(v.toString(), f), se("null", f));
     else static if (is(T == interface))
         handle!(Object)(cast(Object) v, f, sf, se);
-
+    else static if (isForwardRange!T && !isInfinite!T)
+    {
+        Flags old_flag = f.flags;
+        f.flags |= Flags.Nested;
+        sf("[");
```
Comment (AndrejMitrovic) [inline src/ocean/text/convert/Formatter.d:485]: You need to call `.save`. That's how forward ranges work. You essentially write code like this:

```D
auto range = v.save;
while (!range.empty)
{
    range.popFront();
}
```

https://tour.dlang.org/tour/en/basics/ranges
https://dlang.org/phobos/std_range_primitives.html#isForwardRange

Unfortunately the Korean version is outdated and doesn't describe forward ranges on https://tour.dlang.org/tour/kr/basics/ranges.

Comment (MichaelKim20) [reply src/ocean/text/convert/Formatter.d:485]: Well, I finally get the meaning of **"not to consume"** now

File: src/ocean/text/convert/Formatter_test.d [hunk 6]
```diff
@@ -559,3 +559,34 @@ unittest
...
 }
+
+
+/*******************************************************************************
+
+    unit tests for `Range`
+
+*******************************************************************************/
+
+unittest
```
Comment (Geod24) [inline src/ocean/text/convert/Formatter_test.d:570]: Can you add an unittest with:
```
struct Key { int value; }
struct Value { string value; }
Key[Value] aa;
aa.byKeyValue;
```

And some data in it (4 or more elements) ?

File: src/ocean/text/convert/Formatter.d [hunk 7]
```diff
@@ -962,3 +977,163 @@ unittest
     static assert(canSwitchOn!(immutable int));
     static assert(!canSwitchOn!(real));
 }
+
+/// Returns `true` if `R` is an input range.
```
Comment (Geod24) [inline src/ocean/text/convert/Formatter.d]: All of this was extracted from Phobos, right ?

Comment (MichaelKim20) [reply src/ocean/text/convert/Formatter.d]: I referred to Phobos. I made a function that it does not support. And some of the test codes will be similar.
std.range is a phobos's library, so I had to refer to it.

File: src/ocean/text/convert/Formatter_test.d (hunk 6 above)
Comment (MichaelKim20) [reply src/ocean/text/convert/Formatter_test.d:570]: I just added it.

**Review by AndrejMitrovic — APPROVED** [review]

File: src/ocean/text/convert/Formatter.d (hunk 7 above)
Comment (MichaelKim20) [reply src/ocean/text/convert/Formatter.d]: I added a link to the reference code below.
```
    See_Also:
        https://github.com/dlang/phobos/blob/
        418106123d6106efb03b2753bf191215b411ad96/std/range/primitives.d#L124
```

**Comment by MichaelKim20** [comment]
@Geod24 Ready for review

**Review by Geod24 — APPROVED** [review]
